src/Build/Logging/BinaryLogger/BuildEventArgsReader.cs

```diff
@@ -0,0 +1,673 @@
+            int taskId = ReadInt32();
+            int submissionId = ReadInt32();
+            int projectInstanceId = ReadInt32();
+
+            var result = new BuildEventContext(
+                submissionId,
+                nodeId,
+                projectInstanceId,
+                projectContextId,
+                targetId,
+                taskId);
+            return result;
+        }
+
+        private Dictionary<string, string> ReadStringDictionary()
+        {
+            int count = ReadInt32();
+
+            if (count == 0)
+            {
+                return null;
+            }
+
+            Dictionary<string, string> result = new Dictionary<string, string>();
+            for (int i = 0; i < count; i++)
+            {
+                string key = ReadString();
+                string value = ReadString();
+                result[key] = value;
+            }
+
+            return result;
+        }
+
+        private class TaskItem : ITaskItem
+        {
+            public string ItemSpec { get; set; }
+            public Dictionary<string, string> Metadata { get; } = new Dictionary<string, string>();
+
+            public int MetadataCount => Metadata.Count;
+
+            public ICollection MetadataNames => Metadata.Keys;
+
+            public IDictionary CloneCustomMetadata()
+            {
+                return Metadata;
+            }
+
+            public void CopyMetadataTo(ITaskItem destinationItem)
+            {
+                throw new NotImplementedException();
+            }
+
+            public string GetMetadata(string metadataName)
+            {
+                return Metadata[metadataName];
+            }
+
+            public void RemoveMetadata(string metadataName)
+            {
+                throw new NotImplementedException();
+            }
+
+            public void SetMetadata(string metadataName, string metadataValue)
+            {
+                throw new NotImplementedException();
+            }
+        }
+
+        private ITaskItem ReadItem()
+        {
+            var item = new TaskItem();
+            item.ItemSpec = ReadString();
+
+            int count = ReadInt32();
+            for (int i = 0; i < count; i++)
+            {
+                string name = ReadString();
+                string value = ReadString();
+                item.Metadata[name] = value;
+            }
+
+            return item;
+        }
+
+        private IEnumerable ReadItems()
+        {
+            int count = ReadInt32();
+            if (count == 0)
+            {
+                return null;
+            }
+
+            var list = new List<DictionaryEntry>(count);
+
+            for (int i = 0; i < count; i++)
+            {
+                string key = ReadString();
+                ITaskItem item = ReadItem();
+                list.Add(new DictionaryEntry(key, item));
+            }
+
+            return list;
+        }
+
+        private IEnumerable ReadItemList()
```

Comment:
MSBuild has 3 different implementations of ITaskItem that are significant. In one case we just have a list of ITaskItem, in another case it's a dictionary keyed by item type.